credentials/tls/certprovider/store_test.go

```diff
@@ -90,262 +100,297 @@ func (c *fakeStableConfig) Canonical() []byte {
-	// return this on subsequent calls to KeyMaterial().
-	kmCh chan *KeyMaterial
 }
 
-func (p *fakeProvider) run(ctx context.Context) {
-	for {
-		select {
-		case <-ctx.Done():
-			return
-		case km := <-p.kmCh:
-			p.Distributor.Set(km, nil)
-		}
-	}
+// newKeyMaterial allows tests to push new key material to the fake provider
+// which will be made available to users of this provider.
+func (p *fakeProvider) newKeyMaterial(km *KeyMaterial, err error) {
+	p.Distributor.Set(km, err)
 }
 
+// Close helps implement the Provider interface.
 func (p *fakeProvider) Close() {
-	p.cancel()
 	p.Distributor.Stop()
 }
 
 // loadKeyMaterials is a helper to read cert/key files from testdata and convert
-// them into a KeyMaterial struct.
-func loadKeyMaterials() (*KeyMaterial, error) {
-	certs, err := tls.LoadX509KeyPair(testdata.Path("x509/server1_cert.pem"), testdata.Path("x509/server1_key.pem"))
+// them into a KeyMaterialReader struct.
+func loadKeyMaterials(t *testing.T, cert, key, ca string) *KeyMaterial {
+	t.Helper()
+
+	certs, err := tls.LoadX509KeyPair(testdata.Path(cert), testdata.Path(key))
 	if err != nil {
-		return nil, err
+		t.Fatalf("Failed to load keyPair: %v", err)
 	}
 
-	pemData, err := ioutil.ReadFile(testdata.Path("x509/client_ca_cert.pem"))
+	pemData, err := ioutil.ReadFile(testdata.Path(ca))
 	if err != nil {
-		return nil, err
+		t.Fatal(err)
 	}
 	roots := x509.NewCertPool()
 	roots.AppendCertsFromPEM(pemData)
-	return &KeyMaterial{Certs: []tls.Certificate{certs}, Roots: roots}, nil
+	return &KeyMaterial{Certs: []tls.Certificate{certs}, Roots: roots}
 }
 
-func makeProvider(t *testing.T, name, config string) (Provider, *fakeProvider) {
-	t.Helper()
-
-	prov, err := GetProvider(name, config)
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	// The store returns a wrappedProvider, which holds a reference to the
-	// actual provider, which in our case in the fakeProvider.
-	wp := prov.(*wrappedProvider)
-	fp := wp.Provider.(*fakeProvider)
-	return prov, fp
+// kmReader wraps the KeyMaterial method exposed by Provider and Distributor
+// implementations. Defining the interface here makes it possible to use the
+// same helper from both provider and distributor tests.
+type kmReader interface {
+	KeyMaterial(context.Context) (*KeyMaterial, error)
 }
 
-// TestStoreWithSingleProvider creates a single provider through the store and
-// calls methods on it.
-func (s) TestStoreWithSingleProvider(t *testing.T) {
-	prov, fp := makeProvider(t, fakeProvider1Name, fakeConfig)
-
-	// Push key materials into the provider.
-	wantKM, err := loadKeyMaterials()
-	if err != nil {
-		t.Fatal(err)
-	}
-	fp.kmCh <- wantKM
-
-	// Get key materials from the provider and compare it to the ones we pushed
-	// above.
+// readAndVerifyKeyMaterial attempts to read key material from the given
+// provider and compares it against the expected key material.
+func readAndVerifyKeyMaterial(kmr kmReader, wantKM *KeyMaterial) error {
```

Comment:
Accept a context as a parameter?

Reply:
This method is called more than once from few tests, and some of those calls are expected to timeout. So, accepting a context parameter here would mean that the tests would have to redeclare a `ctx` everytime before calling this function.